### PR TITLE
fix: Dashboard aware RBAC dataset permission

### LINKED ISSUE #24782: Gamma user can clone a dashboard and all of its charts without being owner of dashboard, chart and its underlying dataset
Folks, 

Sorry, but I am confused & annoyed with this behavior which I have described below. 

We are using DASHBOARD_RBAC to give read-only dashboard access to users without giving them any access to charts, datasets, and databases. We have assigned only the default Gamma role to the user profile and to the dashboard roles.

see below, SERVICEACCOUNT user is not the owner of the dashboard. The dashboard has the Gamma role assigned.
image

see below, SERVICEACCOUNT user has only Gamma role assigned. 
image

Here are the two issues we are observing:

1) In 2.1.0, 3.0.0rc1, and in the current master, a Gamma user can save a copy of a dashboard not owned by them. More importantly, they can make copies of all charts in the dashboard with the "also copy (duplicate) charts" checkbox. This is very undesired behavior and a maintenance nightmare for us.

see below, SERVICEACCOUNT user can use "save as" on the dashboard and duplicate all charts. 
image

2) They can edit the chart from the dashboard and "save as" a new chart, even though the chart is not owned by them. 

see below, SERVICEACCOUNT user can edit a chart from the dashboard and save it as a new chart from the chart builder.
image

image

see below, SERVICEACCOUNT user is not the owner of the chart.
image

The irony is, Gamma users can't list any charts from the Charts Menu (including cloned charts). An admin can see that charts are being cloned.  

image

They can't see any dataset from the Datasets menu. 

image

They can't see any dataset when trying to create a new chart.

image

This was reported by another user here as well: https://apache-superset.slack.com/archives/CCKHMGRRB/p1688356037634189

This behavior didn't exist in 1.5.3. Is this a deliberately added behavior or is it a bug with DASHBOARD_RBAC or somewhere else in the code?

### Expected results

Gamma users who are not owners of the dashboard shouldn't be able to save a dashboard.
Gamma users who are not owners of the charts shouldn't be able to edit a chart from the dashboard and should not be able to save it as a new chart from the chart builder.

### Actual results

Gamma users can save a dashboard and chart even if they are not owners. 

#### Screenshots

Added above. 

### Environment

- browser type and version: Version 114.0.5735.134 (Official Build) (64-bit)
- superset version: 3.0.0rc1
- python version: 3.9.x
- node.js version: NA
- any feature flags active: This feature flags are set to true ->
ALERT_REPORTS, DASHBOARD_CROSS_FILTERS, DASHBOARD_RBAC, GENERIC_CHART_AXES, ALLOW_FULL_CSV_EXPORT, DRILL_TO_DETAIL, HORIZONTAL_FILTER_BAR

### Checklist

Make sure to follow these steps before submitting your issue - thank you!

- [ ] I have checked the superset logs for Python stack traces and included them here as text if there are any.
- [ x] I have reproduced the issue with at least the latest released version of the superset.
- [x ] I have checked the issue tracker for the same issue and haven't found one similar.

### Additional context

NA

### PR DESCRIPTION
**Pull request by john-bodley**
### SUMMARY

This PR (alongside https://github.com/apache/superset/pull/24806) fixes https://github.com/apache/superset/issues/24782.

Previously users were granted access to a dataset/chart associated with a dashboard (per the RBAC dashboard roles) however it wasn't context aware, i.e., access was granted irrespective of the context of the request. Specifically this PR ensures that role based dashboard access only occurs within the confines of a dashboard.

### TESTING INSTRUCTIONS

Added unit tests.

#### BEFORE

 | 
|-|-|

#### AFTER

 | 
|-|-|

### ADDITIONAL INFORMATION

- [ ] Has associated issue: https://github.com/apache/superset/issues/24782
- [ ] Required feature flags:
- [ ] Changes UI
- [ ] Includes DB Migration (follow approval process in [SIP-59](https://github.com/apache/superset/issues/13351))
  - [ ] Migration is atomic, supports rollback & is backwards-compatible
  - [ ] Confirm DB migration upgrade and downgrade tested
  - [ ] Runtime estimates and downtime expectations provided
- [ ] Introduces new feature or API
- [ ] Removes existing feature or API

cc: @mdeshmu @sfirke

### CODE REVIEW

File: tests/integration_tests/security_tests.py
Comment (john-bodley) [inline tests/integration_tests/security_tests.py:1700]: I'm not sure this is needed, i.e., `test_get_dashboard_view__user_access_with_dashboard_permission` might be suffice.

File: tests/integration_tests/dashboards/security/security_rbac_tests.py
Comment (john-bodley) [inline tests/integration_tests/dashboards/security/security_rbac_tests.py:193]: This test (which was failing after the change) indicates that either i) RBAC were supposed to expend beyond the scope/context of the dashboard, or ii) was a mistake.

File: tests/integration_tests/security/guest_token_security_tests.py
Comment (john-bodley) [inline tests/integration_tests/security/guest_token_security_tests.py:162]: Conditional on my logic being correct, these tests are now superfluous as the irrespective of whether a guest is authorized or not, they don't have access to the chart, dataset, etc. unless it's in the context of RBAC.  The logic previously outlined here condenses to the same logic which is defined in the `security_tests.py`.

**Comment by betodealmeida** [comment]
@john-bodley @Vitor-Avila did a fix in https://github.com/apache/superset/pull/24808 giving access to datasets that power native filters, when using Dashboard RBAC. Can we include his changes here, or would that use case already work with your changes?

**Comment by john-bodley** [comment]
Thanks @betodealmeida for sharing @Vitor-Avila's PR. Initially I can't say whether this change will resolve said issue as it's not apparent where (and with what arguments) the `raise_for_access` method is being invoked from regarding the native filters.

@Vitor-Avila would you mind using my branch to test your issue? Note assuming the approach outlined in this PR is correct (@villebro it's probably the definitive expert on this) and it doesn't address @Vitor-Avila's issue, I do sense the right solution there is make sure that the access for native filters is akin to that of dashboard charts.

Personally I'm a tad worried about the `DASHBOARD_RBAC` feature flag as I don't think all the various paths have been fully flushed out. Furthermore our default security manager/policy is already somewhat of a complex web of rules which only gets worse when you add another variable (the feature flag) to the mix.

**Comment by Vitor-Avila** [comment]
hey @john-bodley, thanks for the quick response! 

My PR was actually not created with the `DASHBOARD_RBAC` FF in mind, but actually focusing on embedded access for dashboards. I'm not sure if the permission flow/logic is the same between guest user/dashboard RBAC access, but I know at least for embedded access you would only specify the dashboard `uuid` when generating the `guest_token`, and I believe the expectation is that the user should be able to view all charts and filters used in the specified dashboard. For embedded dashboards, it's also not possible to access Charts directly (at least not through the iframe) -- but likely through the API.

I'll test if my issue is also addressed with your branch and report back -- thank you! 🙏

**Comment by Vitor-Avila** [comment]
Yup, I confirmed the issue still exists in this branch:
image

@eschutho mentioned you also have a PR to re-introduce the logic from `can_access_based_on_dashboard`: https://github.com/apache/superset/pull/24804

Should I try to re-create my changes on top of this modification?

**Comment by john-bodley** [comment]
Thanks @Vitor-Avila for confirming. I thought that might be the case as the dashboard information is likely lacking from the form-data when the native dashboard filter datasource check is happening. https://github.com/apache/superset/pull/24804 is simply a refactor and thus wouldn't change anything.

The challenge with the RBAC feature is opening up access only in the context of a dashboard. Furthermore I sense there's actually a flaw with the model, i.e., a Gamma user (who has been granted access to a dashboard) could likely have broader access than an owner as it pertains to how access is determined for the datasets within the dashboard which are driving the native filters and charts.

**Comment by Vitor-Avila** [comment]
@john-bodley sorry I mis-read https://github.com/apache/superset/pull/24804 and thought you were re-adding the logic for `can_access_based_on_dashboard` somewhere else, but that's not true - I was confused with the `raise_for_dashboard_access` function. My apologies for the confusion.

In my previous test, I had only checked the issue I'm trying to fix for embedded users, but I just tried with the `DASHBOARD_RBAC` access as well and confirmed it also happens for this flow:
image

Not sure if we can keep the `can_access_based_on_dashboard` function or if that's not compatible with the work you've been doing in these PRs.

**Comment by john-bodley** [comment]
@Vitor-Avila I think we likely need to take a pause on this for now until collectively the community can agree on how the dashboard RBAC are supposed to work. Hopefully this can happen sometime next week once some of the Apache Superset committers who are well versed in this space are back online.

The challenge otherwise is we could find ourselves playing whack-a-mole with trying to fix issues whilst simultaneously creating other ones. Taking a more holistic (30,000') view of problem will likely help everyone better grok the intention and thus what access controls are required.

**Comment by Vitor-Avila** [comment]
sure, that makes sense! It could also be a good opportunity to revisit if both flows (`DASHBOARD_RBAC` and guest user) should be handled in a more distinct way.

**Comment by john-bodley** [comment]
@eschutho I added some comments to https://github.com/apache/superset/pull/24808. I _think_ the underlying logic outlined in said PR is likely correct and thus it can be merged—hopefully after addressing a few small comments.

File: superset/security/manager.py
Comment (villebro) [reply superset/security/manager.py:1848]: Do we need to check that the purported chart whose data is being loaded actually exists on the dashboard?

**Review by villebro — APPROVED** [review]
I think this is taking dashboard RBAC in the right direction, but I think ultimately we'd need some assurance, that the query context that's being requested
1. actually belongs to a specific chart
2. the chart actually belongs to the dashboard that the user has access to

I'm actually probably going to be working on this soon, so I can work on adding more checks later. So LGTM.

**Comment by mdeshmu** [comment]
@michael-s-molina can this be included in 3.0 ?

**Comment by michael-s-molina** [comment]
> @michael-s-molina can this be included in 3.0 ?

@mdeshmu All `fix` PRs that affect 3.0 will be included in the release. I'm cherry-picking them daily. If a `fix` PR is not merged by the time we release 3.0.0, it will be included right after it in 3.0.1. We're planning to submit patch releases every 2 weeks.

File: superset/security/manager.py
Comment (john-bodley) [reply superset/security/manager.py:1848]: @villebro is the logic on line #1846 not suffice, i.e., extracting the dashboard ID from the form-data (which could stem from either a legacy visualization or query-context)?

**Comment by jfrag1** [comment]
@john-bodley I discovered that this PR introduces a security issue that can allow any user to get data from any datasource.  I've opened https://github.com/apache/superset/pull/24996 to revert this, see the PR description there for more details

**Pull request merged**
